Add create attribute term modal

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This allows users to create a new attribute term if the term they are searching for doesn't exist within the attribute term list.
This is currently dependent on #34999 to be merged, which is why it's pointing at that branch.

https://user-images.githubusercontent.com/2240960/196231848-4b686769-c032-42f2-af67-3508b953140c.mp4

Address the term creation part in #34331  .

### How to test the changes in this Pull Request:

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Go to **Products > Add new (MVP)** and fill out some of the initial details like name and price
3. Scroll down and click **Add first attribute** in the attributes section
4. Select one of the attributes
5. Search for a attribute term that doesn't exist yet
6. The 'Create "your new attribute term name"' should show as an option
7. Click that and another modal should show with the create fields.
8. Fill out all the fields and click **Add** this should create the attribute term and add it as a selected item within the terms field
9. With `localStorage.setItem( 'debug', 'wc-admin:*' );` run prior, it should show that a `product_attribute_term_add` and `product_attribute_term_add_success` event has been fired.
10. Open the modal again and select **cancel** this shouldn't do anything.
11. A user should not be able to click **add** when the name is removed.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add create attribute term modal

louwie17 committed:
Add back filter missed during rebase

louwie17 committed:
Add changelog

louwie17 committed:
Fix lint error

joelclimbsthings commented on plugins/woocommerce-admin/client/products/fields/attribute-term-input-field/attribute-term-input-field.tsx:
```diff
@@ -91,100 +100,155 @@ export const AttributeTermInputField: React.FC<
+				getFilteredItems={ ( allItems, inputValue ) => {
+					if (
+						inputValue.length > 0 &&
+						! allItems.find(
```
Just a note that this will fail if it returns an item that is falsy in any way, but it looks unlikely to cause any issues here.

louwie17 replied:
Yeah that should be fine, plus the `find` method already assume the each item contains a `name` attribute, so, so the item being return will have to contain a `name` attribute none the less.

joelclimbsthings commented on plugins/woocommerce-admin/client/products/fields/attribute-term-input-field/attribute-term-input-field.tsx:
```diff
@@ -91,100 +100,155 @@ export const AttributeTermInputField: React.FC<
+													/>
+													<span>
+														{ sprintf(
+															/* translators: The name of the new attribute to be created */
```
Just a note that this is the name of the attribute *term, not attribute, in case it matters to the translators.

louwie17 replied:
Updated

joelclimbsthings commented on plugins/woocommerce-admin/client/products/fields/attribute-term-input-field/create-attribute-term-modal.tsx:
```diff
@@ -0,0 +1,162 @@
+type CreateAttributeTermModalProps = {
+	initialAttributeTermName: string;
+	attributeId: number;
+	onCancel: () => void;
```
I'm wondering if there's a possible use case in which these handlers should be optional?

louwie17 replied:
I don't think so, but I made them optional just incase, I defaulted them to empty functions in this commit: https://github.com/woocommerce/woocommerce/pull/35131/commits/2f39f27c9fc5e41179fc13f8e92f5c3d4a1c6b5d

joelclimbsthings commented on plugins/woocommerce-admin/client/products/fields/attribute-term-input-field/create-attribute-term-modal.tsx:
```diff
@@ -0,0 +1,162 @@
+							<TextControl
+								label={ __( 'Name', 'woocommerce' ) }
+								{ ...nameInputProps }
+								onBlur={ () => {
```
If the `onBlur` here fires, it actually closes both the terms modal and the add attribute modal, which I'd guess wouldn't be expected. Should we stop it from bubbling down to the add modal as well?

louwie17 replied:
Hmm could you maybe share a video of this? Things work as expected for me and the Modal doesn't auto close for me when the `onBlur` is being fired.

joelclimbsthings reviewed (changes requested):
Great job @louwie17 ! 
Added a few very minor comments, and I also noticed that the fields are not responsive on smaller viewports, which might be worth addressing here.

It's also worth noting that we're definitely going to have some conflicts with the [edit attributes PR](https://github.com/woocommerce/woocommerce/pull/35269) which includes some semi-foundational changes, so we might want to get that merged before too long.

louwie17 committed:
Address some feedback from PR review

louwie17 commented:
> Great job @louwie17 ! Added a few very minor comments, and I also noticed that the fields are not responsive on smaller viewports, which might be worth addressing here.

Thanks for the feedback @joelclimbsthings I addressed most of your feedback, the only item I couldn't reproduce was the `onBlur` issue you are encountering, I was curious if you could share a video or something, it seems a bit odd to me that this `onBlur` would trigger that.
I did also address the smaller viewports issue, nice catch!

> It's also worth noting that we're definitely going to have some conflicts with the [edit attributes PR](https://github.com/woocommerce/woocommerce/pull/35269) which includes some semi-foundational changes, so we might want to get that merged before too long.

Get which one merged first?

joelclimbsthings reviewed (changes requested):
Right on, thanks @louwie17 ! I'm not sure if this behavior is actually much of a problem, but I can see it being annoying to users:

![Peek 2022-10-28 12-30](https://user-images.githubusercontent.com/444632/198719512-c1f45b97-0f66-414c-8fdc-38834677d986.gif)

At the end of the animation I'm clicking at the spot I'm circling, and you can see that it's closing _both_ modals, not just the terms modal.

Edit: Either way on merging the PRs. I'm good to resolve the conflicts for this one, but don't want the edit modal PR to get too bogged down with other changes coming down.

louwie17 commented:
> Right on, thanks @louwie17 ! I'm not sure if this behavior is actually much of a problem, but I can see it being annoying to users:
> 
> At the end of the animation I'm clicking at the spot I'm circling, and you can see that it's closing _both_ modals, not just the terms modal.

Aaaha thanks that video helps a lot, I should have a fix Monday morning sometime, it should be relatively simple to handle.

louwie17 committed:
Prevent first modal from closing if closing the second modal when clicking outside

louwie17 commented:
@joelclimbsthings I fixed the issue in this commit: https://github.com/woocommerce/woocommerce/pull/35131/commits/38dcbb319b1f09efd6fae24e2db2e63bfeef124a
Only the second modal should close now, and afterwards the first modal should still close fine.

Hope you have some time to look at it again.

joelclimbsthings commented:
Awesome @louwie17 , works beautifully. 🚢 :shipit:
